spec/integration/github_omniauth_spec.rb

```diff
@@ -0,0 +1,232 @@
+          public_repos: 2,
+          public_gists: 1,
+          followers: 20,
+          following: 0,
+          created_at: "2008-01-14T04:33:35Z",
+          updated_at: "2008-01-14T04:33:35Z",
+          private_gists: 81,
+          total_private_repos: 100,
+          owned_private_repos: 100,
+          disk_usage: 10000,
+          collaborators: 8,
+          two_factor_authentication: true,
+          plan: {
+            name: "Medium",
+            space: 400,
+            private_repos: 20,
+            collaborators: 0
+          }
+        ),
+        headers: {
+          "Content-Type" => "application/json"
+        }
+      )
+  end
+
+  it "doesn't sign in anyone if none of the emails from github are verified" do
+    post "/auth/github"
+
+    setup_github_emails_stub(
+      [
+        {
+          email: user1.email,
+          primary: true,
+          verified: false,
+          visibility: "private"
+        },
+        {
+          email: user2.email,
+          primary: false,
+          verified: false,
+          visibility: "private"
+        }
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
+  end
+
+  it "matches a non-primary email if it's verified and the primary email isn't" do
+    post "/auth/github"
+
+    setup_github_emails_stub(
+      [
+        {
+          email: user1.email,
+          primary: true,
+          verified: false,
+          visibility: "private"
+        },
+        {
+          email: user2.email,
+          primary: false,
+          verified: true,
+          visibility: "private"
+        }
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(
+      UserAuthToken.lookup(request.cookie_jar.encrypted[:_t]["token"]).user.id
+    ).to eq(user2.id)
+  end
+
+  it "doesn't match a non-primary email if it's not verified" do
+    post "/auth/github"
+
+    setup_github_emails_stub(
+      [
+        {
+          email: "somerandomemail@discourse.org",
+          primary: true,
+          verified: true,
+          visibility: "private"
+        },
+        {
+          email: user2.email,
+          primary: false,
+          verified: false,
+          visibility: "private"
+        }
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
+  end
+
+  it "favors the primary email over secondary emails when they're all verified" do
+    post "/auth/github"
```

Comment:
Is there a response we're expecting here?

Reply:
This request redirects to github and sets up some state in the session that are needed for the later request. I've added a couple of assertions for ensuring this requests 302 redirects to github.